### DISCUSSION TITLE
Issues with XAML Behaviors

### DISCUSSION THREAD

**Discussion by majidkhalili** [unoplatform/uno#13494] [Q&A] [answered]
Since UNO Platform doesn't support XAML triggers, I am trying to use the [XAML Behaior](https://github.com/microsoft/XamlBehaviors) with no success. 

Here is the steps I take to create and configure a project and fail:

1. Create a new Blank UNO project.
- Only Windows and Webassembly projects were added.   
- Dot net 7

2. Install [the Nuget package](https://www.nuget.org/packages/Microsoft.Xaml.Behaviors.WinUI.Managed) and add it to all project in the solution. 

At this point, the project builds and runs on both platforms (Windows and web assembly)

3. In the MainPage.xaml: 
3.1. Add the references as below:
```
xmlns:Interactivity="using:Microsoft.Xaml.Interactivity"
xmlns:Core="using:Microsoft.Xaml.Interactions.Core" 
``` 

3.2. Add a button with a basic behavior from the sample code as below:
```
<Button xmlns:Core="using:Microsoft.Xaml.Interactions.Core" xmlns:Interactivity="using:Microsoft.Xaml.Interactivity">
    <Interactivity:Interaction.Behaviors>
        <Core:EventTriggerBehavior EventName="Click">
            <Core:ChangePropertyAction PropertyName="Background">
                <Core:ChangePropertyAction.Value>
                    <SolidColorBrush Color="Red"/>
                </Core:ChangePropertyAction.Value>
            </Core:ChangePropertyAction>
        </Core:EventTriggerBehavior>
    </Interactivity:Interaction.Behaviors>
</Button>
```

At this point I expect the code to be running with the expected behavior. But I get this error:

> Severity	Code	Description	Project	File	Line	Suppression State
> Error	UXAML0001	Could not load type 'Uno.Xaml.IsIncludedResult' from assembly 'Uno.Xaml, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null'.	BehaviorTest4 (net7.0)

I would appreciate it if there are any tips to resolve the issue.

**Comment by limeniye** [comment, marked as answer]
Hello, @majidkhalili.
Thank you for your report. 

To solve this problem check the Uno documentation 
https://platform.uno/docs/articles/uno-community-toolkit.html

and try to use the following steps:

https://github.com/unoplatform/uno/discussions/13192#discussioncomment-6692036

```XML
    <ItemGroup Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">
        <PackageReference Include="CommunityToolkit.WinUI.UI.Behaviors" Version="7.1.2"/>
    </ItemGroup>
    <ItemGroup Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) != 'windows'">
        <PackageReference Include="Uno.CommunityToolkit.WinUI.UI.Behaviors" Version="7.1.200-dev.6.gf657bb2e91"/>
    </ItemGroup>
```

**Reply by michael-hawker** [reply]
We have the new 8.0 version of the Windows Community Toolkit https://devblogs.microsoft.com/ifdef-windows/announcing-windows-community-toolkit-v8-0/ - Which means you can use the single `CommunityToolkit.WinUI.Behaviors` package for your Uno and WinUI 3 projects without any conditional.

**Reply by agneszitte** [reply]
Little update for anyone trying to use XAML Behaviors ([unoplatform/Uno.XamlBehaviors](https://github.com/unoplatform/Uno.XamlBehaviors)) lately with our [latest single project structure](https://platform.uno/blog/the-first-and-only-true-single-project-for-mobile-web-desktop-and-embedded-in-net/), documentation update are coming soon (setup details mentioned in the description) : 
[[Docs] unoplatform/Uno.XamlBehaviors documentation will need to be updated to add details for single project structure](https://github.com/unoplatform/uno/issues/16871#top)
#16871

Documentation update coming soon for Windows Community Toolkit as well (setup details mentioned in the description):
[[Docs] Windows Community Toolkit documentation will need to be updated to add latest details](https://github.com/unoplatform/uno/issues/16872#top)
#16872

**Comment by majidkhalili** [comment]
@limeniye Do you mean I have to use the uno-comunity-toolkit (Uno.CommunityToolkit.WinUI.UI.Behaviors) instead of XAML Behavior (Microsoft.Xaml.Behaviors.WinUI.Managed)?

**Reply by limeniye** [reply]
Exactly. I recommend you to follow the Uno documentation. Its should to fix your problem.

**Reply by majidkhalili** [reply]
Great, I will try it now.

**Reply by limeniye** [reply]
great!

**Reply by majidkhalili** [reply]
I installed the Community Toolkit Bheaviour (CommunityToolkit.WinUI.UI.Behaviors), but couldn't manage to implement a basic trigger action. 

Is there any documentation or sample code for the Behaviors and actions that I can follow? Let's say, I want to change a property (or run an animation) on trigger (e.g. button click) all in XAML code.

**Reply by limeniye** [reply]
Did you install only the ```CommunityToolkit.WinUI.UI.Behaviors``` package, or ```Uno.CommunityToolkit.WinUI.UI.Behaviors``` also?
Please, attached more information of your problem.
The better way, if you will add the sample of your problem (.zip)

**Reply by majidkhalili** [reply]
I just installed the `CommunityToolkit.WinUI.UI.Behaviors`, and since I am using WinUI, I did not install 'Uno.CommunityToolkit.WinUI.UI.Behaviors'. and it is now compiling and running. 

I don't have any sample code yet, because I could not find the documentation to write a trigger and action. I was expecting that behavior works as a trigger, but apparently, it doesn't. I will try it more tonight and send the code here. If you have a sample code with WPF type of triggers in UNO, would be great to share it here.

**Reply by limeniye** [reply]
[UnoBehaviorsSample.zip](https://github.com/unoplatform/uno/files/12538069/UnoBehaviorsSample.zip)

```XML
    xmlns:i="using:Microsoft.Xaml.Interactivity"
    xmlns:core="using:Microsoft.Xaml.Interactions.Core"
```
```XML
<TextBlock x:Name="HelloWorldTextBlock" Text="Hello World!"/>
<CheckBox Content="Hello World Visibility">
    <i:Interaction.Behaviors>
        <core:EventTriggerBehavior EventName="Checked">
            <core:ChangePropertyAction TargetObject="{Binding ElementName=HelloWorldTextBlock}"
                                       PropertyName="Visibility" Value="Visible"/>
        </core:EventTriggerBehavior>
        <core:EventTriggerBehavior EventName="Unchecked">
            <core:ChangePropertyAction TargetObject="{Binding ElementName=HelloWorldTextBlock}"
                                       PropertyName="Visibility" Value="Collapsed"/>
        </core:EventTriggerBehavior>
    </i:Interaction.Behaviors>
</CheckBox>
```

![BehaviorsSample](https://github.com/unoplatform/uno/assets/69314237/5636b8f4-994a-4144-abc4-5b59cdb5ce2c)

**Reply by majidkhalili** [reply]
Thanks a lot, this is exactly what I am looking for and your code com

piles very well on my computer. However, when I tried to create an empty project and follow the steps, I get an error that Page doesn't have a method called InitializeComponent. Which to me sounds like there is a bad referencing. 

I have attached to code here, would be great if you could take a look. Thanks in advance. 

[UNOTest_ComunityToolkit_1.zip](https://github.com/unoplatform/uno/files/12542497/UNOTest_ComunityToolkit_1.zip)

**Reply by limeniye** [reply]
Because in my case I'am using variable package versions.
Lets heck the following files at my solution:

**Directory.Build.Props**
```XML
  <PropertyGroup>
    <CommunityToolkitVersion>7.1.2</CommunityToolkitVersion>
    <UnoCommunityToolkitVersion>7.1.100</UnoCommunityToolkitVersion>
  </PropertyGroup>
```

**Directory.Packages.props**
```XML
<Project ToolsVersion="15.0">
  <ItemGroup>
    <!-- Your package versions -->
    <!-- ............................................ -->

    <!-- My package versions -->
    
    <!-- CommunityToolkit -->
    <PackageVersion Include="CommunityToolkit.WinUI.UI.Behaviors" Version="$(CommunityToolkitVersion)" />
	
    <!-- Uno.CommunityToolkit-->
    <PackageVersion Include="Uno.CommunityToolkit.WinUI.UI.Behaviors" Version="$(UnoCommunityToolkitVersion)" />
  </ItemGroup>
</Project>
```

**Reply by majidkhalili** [reply]
Thats exactly it. Thanks a lot.

**Comment by KWodarczyk** [comment]
I have issue with xaml Behaviors. I have created new project using latest uno templates, added one package Uno.Microsoft.Xaml.Behaviors.WinUI.Managed as described here https://github.com/unoplatform/Uno.XamlBehaviors

As soon as i add 

`      xmlns:Interactivity="using:Microsoft.Xaml.Interactivity"
      xmlns:Interactions="using:Microsoft.Xaml.Interactions.Core"`

to my page i get error below and can't build:

Severity	Code	Description	Project	File	Line	Suppression State	Details
Error (active)	WMC1006	Cannot resolve Assembly or Windows Metadata file 'Type universe cannot resolve assembly: Uno.UI, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null.'	UnoWithXamlBehaviours (net8.0-windows10.0.19041)	C:\Users\Administrator\Repos\UnoBugs\UnoWithXamlBehaviours\UnoWithXamlBehaviours\UnoWithXamlBehaviours.csproj	1	

I

**Reply by limeniye** [reply]
@KWodarczyk Hi.
That is so old implementation.
Try to use this:
https://www.nuget.org/packages/CommunityToolkit.WinUI.Behaviors/

**Reply by KWodarczyk** [reply]
I will but why this is not documented anywhere?

**Reply by limeniye** [reply]
Previously, the Behaviors documentation described how to integrate this package, as it was distributed into two separate packages:
> ```xml
> 
> <PackageVersion Include="CommunityToolkit.WinUI.UI.Behaviors" Version="$(CommunityToolkitVersion)" />
> 
> <PackageVersion Include="Uno.CommunityToolkit.WinUI.UI.Behaviors" Version="$(UnoCommunityToolkitVersion)" />
> ```
Currently, there is no need for this:
> ```xml
> <PackageVersion Include="CommunityToolkit.WinUI.Behaviors" />
> ```
 so the documentation describes the general format for integrating CommunityToolkit libraries:
https://platform.uno/docs/articles/uno-community-toolkit-v8.html?tabs=singleproj#referencing-the-windows-community-toolkit

Attached issue: https://github.com/unoplatform/uno/issues/16872

**Comment by KWodarczyk** [comment]
@limeniye  thank you i will try to use Communitytoolkit.WinUI.Behaviors but does it include Behaviors from https://github.com/Microsoft/XamlBehaviors/wiki ?

**Reply by limeniye** [reply]
You can check the list of Behaviors or other information here:
https://learn.microsoft.com/en-us/dotnet/communitytoolkit/windows/behaviors/
https://github.com/CommunityToolkit/Windows/blob/main/components/Behaviors/src/ReadMe.md

**Comment by KWodarczyk** [comment]
@limeniye  sorry but im still confused. I am looking at this:
https://github.com/CommunityToolkit/Windows/blob/main/components/Behaviors/samples/InvokeActionsSample.xaml

In the page namespace declarations i see:

`      xmlns:behaviors="using:CommunityToolkit.WinUI.Behaviors"
      xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
      xmlns:interactions="using:Microsoft.Xaml.Interactions.Core"
      xmlns:interactivity="using:Microsoft.Xaml.Interactivity"`

does that mean in my Uno App i also need to include 

      xmlns:interactions="using:Microsoft.Xaml.Interactions.Core"
      xmlns:interactivity="using:Microsoft.Xaml.Interactivity
?

Is there any example on how to use Behaviors from CommunityToolkit in Uno app ?

**Reply by limeniye** [reply]
You do not need to add additional 'Microsoft.Xaml' packages to your project. Everything needed for Behaviors is already included in the CommunityToolkit.WinUI.Behaviors package.
But do you need to declare these two namespaces -- yes, if your task requires it.

> ```XML 
> xmlns:interactions="using:Microsoft.Xaml.Interactions.Core"
> xmlns:interactivity="using:Microsoft.Xaml.Interactivity
> ```

But when I tried to do this, I encountered the following unpleasant situation:

![image](https://github.com/user-attachments/assets/277f8182-edb3-439b-ab30-c20c57fbd308)

cc @michael-hawker
[CommunityToolkit-Behaviors-NotExist.zip](https://github.com/user-attachments/files/17245362/CommunityToolkit-Behaviors-NotExist.zip)
